### PR TITLE
[Object Spilling] Handle IO worker failures correctly

### LINKED ISSUE #11096: [Object Spilling] Restart spilling workers when crasehd.
### Describe your feature request

If spawned spilling workers are killed or crashed, object spilling doesn't work as expected.

EX)
```python3
def test_spill_objects_manually(shutdown_only):
    # Limit our object store to 75 MiB of memory.
    ray.init(
        object_store_memory=75 * 1024 * 1024,
        _object_spilling_config={
            "type": "filesystem",
            "params": {
                "directory_path": "/tmp"
            }
        },
        _system_config={
            "object_store_full_max_retries": 0,
            "max_io_workers": 4,
        })
    arr = np.random.rand(1024 * 1024)  # 8 MB data
    replay_buffer = []
    pinned_objects = set()
    spilled_objects = set()
    import time

    # Create objects of more than 200 MiB.
    for _ in range(25):
        ref = None
        while ref is None:
            try:
                ref = ray.put(arr)
                replay_buffer.append(ref)
                pinned_objects.add(ref)
            except ray.exceptions.ObjectStoreFullError:
                ref_to_spill = pinned_objects.pop()
                ray.experimental.force_spill_objects([ref_to_spill])
                spilled_objects.add(ref_to_spill)
    time.sleep(30)
    # Kill spilling workers here.

    # Spill 2 more objects so we will always have enough space for
    # restoring objects back.
    refs_to_spill = (pinned_objects.pop(), pinned_objects.pop())
    ray.experimental.force_spill_objects(refs_to_spill)
    spilled_objects.update(refs_to_spill)

    # randomly sample objects
    for _ in range(100):
        ref = random.choice(replay_buffer)
        if ref in spilled_objects:
            ray.experimental.force_restore_spilled_objects([ref])
        sample = ray.get(ref)
        assert np.array_equal(sample, arr)
```

### PR DESCRIPTION
**Pull request by kfstorm**
## Why are these changes needed?

Currently, when a spill/restore worker fails and the state of it in the worker pool is idle, the worker pool will not clean up the metadata of the worker. Subsequent spill/restore requests will reuse this dead worker and RPC requests cannot succeed. This results in broken object spilling functionality.

This PR addresses the issue by removing disconnected IO workers from `registered_io_workers` and `idle_io_workers`.

## Related issue number

Closes #11096

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by kfstorm** [comment]
@rkooo567 As you can see from the CI results, the added test case cannot pass on mac. psutil cannot get the title of processes on mac. Do you have an alternative way to verify that spill workers are running?

File: src/ray/raylet/worker_pool.cc [hunk 1]
```diff
@@ -770,10 +774,16 @@ void WorkerPool::PopIOWorkerInternal(
     // We must fill the pending task first, because 'TryStartIOWorkers' will
     // start I/O workers according to the number of pending tasks.
     io_worker_state.pending_io_tasks.push(callback);
+    RAY_LOG(DEBUG) << "Failed to pop an IO worker. Try starting a new one. Worker type: "
```
Comment (MissiontoMars) [inline src/ray/raylet/worker_pool.cc:777]: Is "There are no idle workers, try starting a new one." sounds better?

**Comment by rkooo567** [comment]
@kfstorm I think you can obtain the name from cmdline or sth instead of name? Did you try this? I believe we have plenty of psutil usage to get proc name in our tests

File: src/ray/raylet/worker_pool_test.cc [hunk 2]
```diff
@@ -1538,11 +1538,83 @@ TEST_F(WorkerPoolTest, PopWorkerStatus) {
...
+  RAY_CHECK_OK(worker_pool_->RegisterWorker(worker2, proc2.GetId(), proc2.GetId(),
+                                            worker_pool_->GetStartupToken(proc2),
+                                            [](Status, int) {}));
+  worker_pool_->OnWorkerStarted(worker2);
+  worker_pool_->PushSpillWorker(worker2);
+  ASSERT_EQ(spill_worker_set.size(), 1);
+  ASSERT_EQ(worker2, *spill_worker_set.begin());
+  // The popped spill worker should be newly created.
+  ASSERT_FALSE(worker_ids.count(worker2->WorkerId()));
+}
```
Comment (rkooo567) [inline src/ray/raylet/worker_pool_test.cc:1607]: Can we add one more test test to start 4 Popworkers, failing all of them, and check all internal states are internal states are properly cleaned up?

**Review by rkooo567 — COMMENTED** [review]
LGTM. Just one test suggestion.

File: src/ray/raylet/worker_pool_test.cc (hunk 2 above)
Comment (kfstorm) [reply src/ray/raylet/worker_pool_test.cc:1607]: I don't quite get your point. `MAX_IO_WORKER_SIZE` is 2 in the worker pool test. Why do you need to pop worker 4 times? Do you try to test the case IO workers fail to start up?

**Comment by kfstorm** [comment]
> @kfstorm I think you can obtain the name from cmdline or sth instead of name? Did you try this? I believe we have plenty of psutil usage to get proc name in our tests

You are right. I've updated the test code. But in my understanding, some test cases won't work due to this different behavior between macOS and Linux. e.g. This `wait_for_condition` will always pass immediately on macOS because `all_workers_exited` always returns true:

https://github.com/ray-project/ray/blob/b3a9d4d87d31255e0871852efaf1ddff7acdc390/python/ray/tests/test_multi_node.py#L69-L79

File: src/ray/raylet/worker_pool_test.cc (hunk 2 above)
Comment (rkooo567) [reply src/ray/raylet/worker_pool_test.cc:1607]: I don't remember why I said 4 lol. But I wanted to see if the internal states are cleaned up on the edge case (PopWorker keeps failing)

**Comment by rkooo567** [comment]
> You are right. I've updated the test code. But in my understanding, some test cases won't work due to this different behavior between macOS and Linux. e.g. This wait_for_condition will always pass immediately on macOS because all_workers_exited always returns true:

Oh, wow this is very bad. Can you create an issue?

File: src/ray/raylet/worker_pool_test.cc (hunk 2 above)
Comment (rkooo567) [reply src/ray/raylet/worker_pool_test.cc:1607]: Lmk if this makes sense. Put the author action required label until then

**Comment by kfstorm** [comment]
@rkooo567 We just found another edge case. If an IO worker is crashed while in use and then pushed back into the worker pool, we should ignore the pushing. Check this out: 325d7ad

File: src/ray/raylet/worker_pool_test.cc (hunk 2 above)
Comment (kfstorm) [reply src/ray/raylet/worker_pool_test.cc:1661]: @rkooo567 I looked at the code of `PopIOWorkerInternal` and it seems that there's no way to test a pop failure as long as we don't change the code of worker pool just for testing. A worker process will always be started. We can make the started mock process fail to register, though, but that requires something like `time.sleep` in the test case.

**Comment by kfstorm** [comment]
> > You are right. I've updated the test code. But in my understanding, some test cases won't work due to this different behavior between macOS and Linux. e.g. This wait_for_condition will always pass immediately on macOS because all_workers_exited always returns true:
> 
> Oh, wow this is very bad. Can you create an issue?

@rkooo567 Issue created. https://github.com/ray-project/ray/issues/20982

File: src/ray/raylet/worker_pool_test.cc (hunk 2 above)
Comment (rkooo567) [reply src/ray/raylet/worker_pool_test.cc:1661]: isn't 
```
PopIOWorker()
worker = CreateWorker()
DisconnectWorker(worker)
```

the scenario where PopWorker is failed? Or am I missing something?

File: src/ray/raylet/worker_pool.cc [hunk 3]
```diff
@@ -1186,8 +1204,18 @@ void WorkerPool::PrestartWorkers(const TaskSpecification &task_spec, int64_t bac
 
 bool WorkerPool::DisconnectWorker(const std::shared_ptr<WorkerInterface> &worker,
                                   rpc::WorkerExitType disconnect_type) {
+  MarkPortAsFree(worker->AssignedPort());
```
Comment (scv119) [inline src/ray/raylet/worker_pool.cc:1207]: hmm would this cause any problem, such as race condition?

**Review by scv119 — APPROVED** [review]

**Review by scv119 — DISMISSED** [review]
oops, not plan to accept it!

**Comment by rkooo567** [comment]
Also ping for https://github.com/ray-project/ray/pull/20752#discussion_r766790227

**Comment by kfstorm** [comment]
@rkooo567 Sorry that I still don't get your point. And I don't have the bandwidth to continue this PR. Maybe you can write the test case and send me a patch?

**Comment by rkooo567** [comment]
@jiaodong @stephanie-wang this is one of known issues in object spilling.

**Comment by rkooo567** [comment]
Maybe we should take it over

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by kfstorm** [comment]
@rkooo567 How about we merge this PR first?

**Comment by rkooo567** [comment]
Hmm yeah I am fine with that.

cc @scv119 can you take a look?

**Comment by rkooo567** [comment]
@kfstorm can you fix tests?

**Review by rkooo567 — APPROVED** [review]

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```
